File: src/main/java/algo/sort/PancakeSorting_969.java
package algo.sort;

import java.util.ArrayList;
import java.util.List;

public class PancakeSorting_969 {
    public static List<Integer> pancakeSort(int[] arr) {
        var len = arr.length;
        var result = new ArrayList<Integer>();
        // swap the max to the end
        var max = len;
        for (var i = len - 1; i > 0; i--) {
            for (var j = 0; j < i; j++) {
                if (arr[j] == max) {
                    reverse(arr, j);
                    result.add(j + 1);
                    reverse(arr, i);
                    result.add(i + 1);
                    break;
                }
            }
            max--;
        }

        return result;
    }

    private static void reverse(int[] arr, int pos) {
        int tmp;
        for (var i = 0; i <= pos / 2; i++) {
            tmp = arr[i];
            arr[i] = arr[pos - i];
            arr[pos - i] = tmp;
        }
    }

    public static void main(String[] args) {
        System.out.println(pancakeSort(new int[]{3,2,4,1}));
//        System.out.println(pancakeSort(new int[]{1,2,3}));
//        System.out.println(pancakeSort(new int[]{2,1,3}));
    }
}

File: src/main/kotlin/algo/sort/PancakeSorting_969.kt
package algo.sort

class KtPancakeSorting_969 {
    fun pancakeSort(arr: IntArray): List<Int> {
        // for [0, i], swap the max val to the end
        var max = arr.size
        val result: MutableList<Int> = mutableListOf()
        for (i in arr.size - 1 downTo 1) {
            for (j in 0..<i) {
                // swap max to head if it is not the last one
                if (arr[j] == max) {
                    reverse(arr, j)
                    result.add(j + 1)

                    // swap max to end
                    reverse(arr, i)
                    result.add(i + 1)
                    break
                }
            }
            max--
        }
        return result
    }

    private fun reverse(arr: IntArray, end: Int) {
        var left = 0; var right = end
        while (left < right) {
            val tmp = arr[left]
            arr[left] = arr[right]
            arr[right] = tmp
            left++
            right--
        }
    }
}

fun main() {
    val pancake = KtPancakeSorting_969()
    println(pancake.pancakeSort(intArrayOf(3,2,4,1)))
}